hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreFileWriter.java

```diff
@@ -256,156 +234,571 @@ public void appendMobMetadata(SetMultimap<TableName, String> mobRefSet) throws I
   public void appendTrackedTimestampsToMetadata() throws IOException {
     // TODO: The StoreFileReader always converts the byte[] to TimeRange
     // via TimeRangeTracker, so we should write the serialization data of TimeRange directly.
-    appendFileInfo(TIMERANGE_KEY, TimeRangeTracker.toByteArray(timeRangeTracker));
-    appendFileInfo(EARLIEST_PUT_TS, Bytes.toBytes(earliestPutTs));
+    liveFileWriter.appendTrackedTimestampsToMetadata();
+    if (historicalFileWriter != null) {
+      historicalFileWriter.appendTrackedTimestampsToMetadata();
+    }
   }
 
   /**
    * Record the earlest Put timestamp. If the timeRangeTracker is not set, update TimeRangeTracker
    * to include the timestamp of this key
    */
   public void trackTimestamps(final Cell cell) {
-    if (KeyValue.Type.Put.getCode() == cell.getTypeByte()) {
-      earliestPutTs = Math.min(earliestPutTs, cell.getTimestamp());
+    liveFileWriter.trackTimestamps(cell);
```

Comment:
I do not think this is the correct behavior, we should only call the write's trackTimestamps method if the cell has been written to the writer...

Reply:
The delete code here is move to SingleStoreFileWriter#trackTimestamps(). So the behavior did not change. 
`    private void trackTimestamps(final Cell cell) {
      if (KeyValue.Type.Put.getCode() == cell.getTypeByte()) {
        earliestPutTs = Math.min(earliestPutTs, cell.getTimestamp());
      }
      timeRangeTracker.includeTimestamp(cell);
    }`